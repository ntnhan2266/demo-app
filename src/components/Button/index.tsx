import React from 'react';
import classNames from 'classnames';

interface ButtonProps {
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  label: string;
  type?: 'button' | 'submit' | 'reset';
  variant?: 'primary' | 'secondary' | 'success' | 'danger';
  disabled?: boolean;
  fullWidth?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  onClick,
  label,
  type = 'button',
  variant = 'primary',
  disabled = false,
  fullWidth = false,
}): React.ReactElement => {
  const buttonClasses = classNames(
    'px-4 py-2 rounded-md focus:outline-none',
    {
      'bg-blue-500 text-white': variant === 'primary',
      'bg-gray-500 text-white': variant === 'secondary',
      'bg-green-500 text-white': variant === 'success',
      'bg-red-500 text-white': variant === 'danger',
      'cursor-not-allowed opacity-50': disabled,
      'w-full': fullWidth
    }
  );

  return (
    <button
      onClick={onClick}
      type={type}
      className={buttonClasses}
      disabled={disabled}
    >
      {label}
    </button>
  );
};

export default Button;
