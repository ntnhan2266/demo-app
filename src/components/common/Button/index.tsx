import React, { ReactNode } from 'react';
import classNames from 'classnames';

interface IProps extends React.InputHTMLAttributes<HTMLButtonElement> {
  label: string | ReactNode;
  variant?: 'primary' | 'secondary' | 'success' | 'danger' | 'ghost';
  fullWidth?: boolean;
  type?: 'button' | 'reset' | 'submit';
}

const Button: React.FC<IProps> = ({
  onClick,
  label,
  variant = 'primary',
  disabled = false,
  fullWidth = false,
  type = 'button',
  className = '',
  ...props
}): React.ReactElement => {
  const buttonClasses = classNames(`px-4 py-2 rounded-md focus:outline-none ${className}`, {
    'bg-blue-500 text-white': variant === 'primary',
    'bg-gray-500 text-white': variant === 'secondary',
    'bg-green-500 text-white': variant === 'success',
    'bg-red-500 text-white': variant === 'danger',
    'bg-transparent text-gray-600 border border-gray-500 hover:bg-gray-200': variant === 'ghost',
    'cursor-not-allowed opacity-50': disabled,
    'w-full': fullWidth,
  });

  return (
    <button type={type} onClick={onClick} className={buttonClasses} disabled={disabled} {...props}>
      {label}
    </button>
  );
};

export default Button;
