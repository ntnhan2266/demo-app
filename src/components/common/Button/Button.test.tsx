import { render, screen, fireEvent } from '@testing-library/react';
import Button from '@/components/common/Button';
import renderer from 'react-test-renderer';

describe('Button component', () => {
  test('should match the snapshot', () => {
    const tree = renderer.create(<Button label='OK' />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders button with label', () => {
    render(<Button label='Click me' />);
    const button = screen.getByText('Click me');
    expect(button).toBeInTheDocument();
  });

  test('handles onClick event', () => {
    const onClickMock = jest.fn();
    render(<Button label='Click me' onClick={onClickMock} />);
    const button = screen.getByText('Click me');
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalled();
  });

  test('applies primary variant styles', () => {
    render(<Button label='Primary' variant='primary' />);
    const button = screen.getByText('Primary');
    expect(button).toHaveClass('bg-indigo-500');
  });

  test('applies secondary variant styles', () => {
    render(<Button label='Secondary' variant='secondary' />);
    const button = screen.getByText('Secondary');
    expect(button).toHaveClass('bg-gray-500');
  });

  test('applies success variant styles', () => {
    render(<Button label='Success' variant='success' />);
    const button = screen.getByText('Success');
    expect(button).toHaveClass('bg-green-500');
  });

  test('applies danger variant styles', () => {
    render(<Button label='Danger' variant='danger' />);
    const button = screen.getByText('Danger');
    expect(button).toHaveClass('bg-red-500');
  });

  test('applies ghost variant styles', () => {
    render(<Button label='Ghost' variant='ghost' />);
    const button = screen.getByText('Ghost');
    expect(button).toHaveClass('text-white');
    expect(button).toHaveClass('focus:outline-none');
  });

  test('applies fullWidth styles', () => {
    render(<Button label='Full Width' fullWidth />);
    const button = screen.getByText('Full Width');
    expect(button).toHaveClass('w-full');
  });

  test('applies disabled styles', () => {
    render(<Button label='Disabled' disabled />);
    const button = screen.getByText('Disabled');
    expect(button).toHaveAttribute('disabled');
    expect(button).toHaveClass('cursor-not-allowed');
    expect(button).toHaveClass('opacity-50');
  });
});
