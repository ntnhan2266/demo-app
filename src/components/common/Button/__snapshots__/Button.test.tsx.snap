// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should match the snapshot 1`] = `
<button
  className="rounded-md focus:outline-none text-white text-base px-4 py-2 bg-indigo-500 hover:bg-indigo-600"
  disabled={false}
  type="button"
>
  OK
</button>
`;
